sw/device/silicon_owner/bare_metal/BUILD

```diff
@@ -230,5 +230,8 @@ opentitan_functest(
     key_struct = "multislot",
     ot_flash_binary = ":rom_ext_virtual_ottf_bl0_virtual",
     signed = True,
-    targets = ["cw310_rom_with_fake_keys"],
+    targets = [
+        "cw310_rom_with_fake_keys",
+        "qemu_cw310_rom_with_fake_keys",
```

Comment:
I don't particularly care for this name.  I think we should think about qemu as a separate execution platfrom than the cw310, even if they share some low-level details like clock configurations and uart dividers.

Reply:
I tend to agree and this is how I initially did it. But the QEMU folks really seem to want to use the cw310 binaries. @rivos-eblot can hopefully explain why?

Reply:
The very goal of QEMU (OT version but any upstream version as well) is to emulate _real_ devices, using the very same, unmodified binaries that run on the real HW. In that sense, QEMU is not a specific target, it emulates existing ones.

In other words, if we had some spare time (🤣), QEMU would be able to execute code for not only CW310, but also CW305 or any other target, not one dedicated to QEMU.

Reply:
I suppose this is being a bit pedantic, but the purpose of the FPGA boards is to emulate the real chip, so we're emulating an emulation :)

Are there parameters that specify the speeds of the various peripheral domains so that the various peripherals with clock dividers do "the right thing" when firmware programs a certain value in the divider? (e.g. UART_NCO).  My previous experience with qemu is that low-speed serial interfaces generally don't implement or simulate the speed of the device and usually implement something like a socket or file for external interactions with the target device.

Without emulating the clock ratio differences between the peripheral domains, is there a meaningful distinction between CW{305, 310, 340} and the real chip as far as qemu is concerned?

Reply:
> I suppose this is being a bit pedantic, but the purpose of the FPGA boards is to emulate the real chip, so we're emulating an emulation :)

True, but when the real chip is available, the goal stays the same: being able to run the binaries built for the chip on QEMU, not built for a "QEMU chip". FPGA does not emulate the real chip, it emulates a variation from it. It is easier to think in term of binary compatibility: QEMU should be able to run the unmodified SW for the selected target.

> My previous experience with qemu is that low-speed serial interfaces generally don't implement or simulate the speed of the device and usually implement something like a socket or file for external interactions with the target device.

True.

> Are there parameters that specify the speeds of the various peripheral domains so that the various peripherals with clock dividers do "the right thing" when firmware programs a certain value in the divider? (e.g. UART_NCO). 

UART clock is simply ignored for now, as the virtual UART is usually only a standard output, a TCP socket on the host, etc. It could be used to compute the real bitrate when QEMU is connected to a real host serial device, but this is not really supported for now. That said, the UART is most an exception for clock management. For example, timers are configured in OT SW for - for instance - the FPGA peripheral clock. QEMU needs to know this peripheral clock value so when the Timer prescaler is set by the OT SW, it can configure its own virtual timer with the right values to generate the same (virtual) ticks.

> Without emulating the clock ratio differences between the peripheral domains, is there a meaningful distinction between CW{305, 310, 340} and the real chip as far as qemu is concerned?

I never look at the differences between the various emulations, but as far as the very same set of IPs, memory map, IRQ, I/Os, clocks etc are the same, there should be no difference. Which means that the very same ELF binaries should run without any change on all these platforms as well. Clock values for all time-related features. That would start with the Timer and AonTimer, but could extend to other devices depending on what is emulated and how it is emulated.

Reply:
Side notes:
1. We initially chose the CE310 as this was the only reference implementation we could use to compare the behavior of QEMU vs a physical device
2. Once the EarlGrey chip is available, it would be nice to upgrade the QEMU VM to support the real chip rather than its FPGA variant (or create another virtual machine)

Reply:
I think we can revisit the exact name when we start converting the existing build/test rules over to the new rules ( issue #19493).  The new rules make creating a new execution environment (e.g. qemu) that derives from an existing environment easier - the qemu environment can inherit the parameters used by the cw310 environment.